firestore/bulkwriter.go

```diff
@@ -0,0 +1,325 @@
+		}
+		return writeResultFromProto(bwr.result)
+	}
+}
+
+// setError ensures that an error is returned on the error channel of BulkWriterJob.
+func (j *BulkWriterJob) setError(e error) {
+	bwr := bulkWriterResult{
+		err:    e,
+		result: nil,
+	}
+	j.resultChan <- bwr
+	close(j.resultChan)
+}
+
+// A BulkWriter supports concurrent writes to multiple documents. The BulkWriter
+// submits document writes in maximum batches of 20 writes per request. Each
+// request can contain many different document writes: create, delete, update,
+// and set are all supported.
+//
+// Only one operation (create, set, update, delete) per document is allowed.
+// BulkWriter cannot promise atomicity: individual writes can fail or succeed
+// independent of each other. Bulkwriter does not apply writes in any set order;
+// thus a document can't have set on it immediately after creation.
+type BulkWriter struct {
+	database        string           // the database as resource name: projects/[PROJECT]/databases/[DATABASE]
+	start           time.Time        // when this BulkWriter was started; used to calculate qps and rate increases
+	vc              *vkit.Client     // internal client
+	maxOpsPerSecond int              // number of requests that can be sent per second
+	docUpdatePaths  map[string]bool  // document paths with corresponding writes in the queue
+	limiter         rate.Limiter     // limit requests to server to <= 500 qps
+	bundler         *bundler.Bundler // handle bundling up writes to Firestore
+	ctx             context.Context  // context for canceling all BulkWriter operations
+	isOpenLock      sync.Mutex       // guards against setting isOpen concurrently
+	isOpen          bool             // flag that the BulkWriter is closed
+}
+
+// newBulkWriter creates a new instance of the BulkWriter.
+func newBulkWriter(ctx context.Context, c *Client, database string) *BulkWriter {
+	// Although typically we shouldn't store Context objects, in this case we
+	// need to pass this Context through to the Bundler handler.
+	ctx = withResourceHeader(ctx, c.path())
+
+	bw := &BulkWriter{
+		database:        database,
+		start:           time.Now(),
+		vc:              c.c,
+		isOpen:          true,
+		maxOpsPerSecond: defaultStartingMaximumOpsPerSecond,
+		docUpdatePaths:  make(map[string]bool),
+		ctx:             ctx,
+		limiter:         *rate.NewLimiter(rate.Limit(maxWritesPerSecond), 1),
+	}
+
+	// can't initialize within struct above; need instance reference to BulkWriter.send()
+	bw.bundler = bundler.NewBundler(&BulkWriterJob{}, bw.send)
+	bw.bundler.HandlerLimit = bw.maxOpsPerSecond
+	bw.bundler.BundleCountThreshold = maxBatchSize
+
+	return bw
+}
+
+// End sends all enqueued writes in parallel and closes the BulkWriter to new requests.
+// After calling End(), calling any additional method automatically returns
+// with an error. This method completes when there are no more pending writes
+// in the queue.
+func (bw *BulkWriter) End() {
```

Comment:
I think there might be a minor race condition here. Here's the idea:

* Goroutine 1 in `checkWriteCondition` acquires the lock, sees that `isOpen` is true, and then releases the lock
* Goroutine 2 in `End` acquires the lock, sets `isOpen` to false, releases the lock, and then flushes
* Goroutine 1 meanwhile starts a write

The question is: does the write success or not in this case?

One possible option would be to use a read write mutex, where the read mutex wraps the entire write attempt and the write mutex wraps `isOpen` here in `End`.

Reply:
Very good point.

Here's what I've done to address this:

- Changed `BulkWriter.isOpenLock` to a [`RWMutex`](https://pkg.go.dev/sync#RWMutex).
- Added a `Lock()` and deferred `Unlock()` at the top of each DB mutator/write function (`Create()`, `Delete()`, etc).

Each mutator will hold the lock while queueing up the write for the Bundler. If `End()` is called during this time, it will will block while the mutator holds the lock.